## Authors: Francisco J. Romero-Campero
##          Ana Bel√©n Romero-Losada
## Contact: Francisco J. Romero-Campero - email: fran@us.es

args <- commandArgs(trailingOnly=TRUE)

output.file <- args[1] #"/home/fran/tmp/chip_test/results/test.Rmd"
microalgae <- args[2]  #"Micromonas pusilla CCMP1545"
mode <- args[3] #"histone_modification"
name <- args[4] #"H3K4me3"
number.replicates <- as.numeric(args[5]) #2
included.control <- args[6] #"yes"

if(mode == "histone_modification")
{
  mode.header <- "Histone Modification"
  mode.intro <- "histone modification"
} else if (mode == "transcription_factor")
{
  mode.header <- "Transcription Factor"
}

## Document header
header <- paste0(paste(c("# **ChIP-seq Data Analysis for the",
                         mode.header, name,
                         "in the microalgae",
                         microalgae),collapse=" "),
                 "**")

## Introduction test
write(x = header,file = output.file,append = F)
write(x = "## Experimental Design and Pre-processing",file = output.file,append = T)

intro.line <- paste(c("This report was automatically generated by **MARACAS 
                      (MicroAlgae RNA-seq and ChIP-seq Analysis)** as an 
                      identification of the genomic loci occupied by the ",
                      mode.intro, " ",name, "in the microalgae **",
                      microalgae, "**. The experimental design considers ", 
                      number.replicates, " replicates"), collapse="")

if(included.control == "yes")
{
  intro.line <- paste0(intro.line, " and control samples. ")
} else
{
  intro.line <- paste0(intro.line,". ")
}

intro.line <- paste0(intro.line, "A summary of the processing of each individual
sample is presented below. Click on the corresponding links to 
access a **quality control analysis** for the raw sequencing data of 
each sample:\n")

write(x = intro.line,file = output.file,append = T)

## Experimental design
write(x = paste(c("* **ChIP samples:**"),collapse=""),file = output.file,
      append = T)

for(i in 1:number.replicates)
{
  write(x = paste(c("\t * **ChIP replicate",
                    i, "**:"),collapse=""),file = output.file,
        append = T)
  write(x=paste(c("[Quality Control analysis](../replicates/replicate_",i,
                  "/chip_1_fastqc.html) "),collapse=""), 
        file=output.file, append=T)
  
  mapping.stats <- readLines(con = paste(c( "../replicates/replicate_",i,
                                            "/chip_mapping_stats_",i),collapse=""),n = 6) 
  write(x = "\n",file=output.file, append=T)
  for(j in 1:6)
  {
    write(x=paste(c("\t\t\t",mapping.stats[j],"\n"),collapse = ""),file=output.file, append=T)
  }
  write(x = "\n",file=output.file, append=T)
}

write(x = paste(c("* **Control (Input or Mock) samples:**"),collapse=""),file = output.file,
      append = T)

for(i in 1:number.replicates)
{
  write(x = paste(c("\t * **Control replicate",
                    i, "**:"),collapse=""),file = output.file,
        append = T)
  write(x=paste(c("[Quality Control analysis](../replicates/replicate_",i,
                  "/control_1_fastqc.html) "),collapse=""), 
        file=output.file, append=T)
  
  mapping.stats <- readLines(con = paste(c( "../replicates/replicate_",i,
                                            "/control_mapping_stats_",i),collapse=""),n = 6) 
  write(x = "\n",file=output.file, append=T)
  for(j in 1:6)
  {
    write(x=paste(c("\t\t\t",mapping.stats[j],"\n"),collapse = ""),file=output.file, append=T)
  }
  write(x = "\n",file=output.file, append=T)
}

write(x = "\n", file=output.file, append=T)

write(x = paste(c("## Identification of the Genomic loci occupied by the ",
                  mode.intro, " ", name, "."),
                collapse=""), file = output.file, append = T)

write(x = "\n", file=output.file, append=T)
write(x=paste(c("The genomic loci occupied by", name,  "were determined using 
      [macs2](https://pypi.org/project/MACS2/). Click on the links
      below to access for each replicate the complete macs2 logs, the mapping signal
      in bigWig format and the identified occupied genomic loci:\n"),collapse=" "), file=output.file, append=T)

for(i in 1:number.replicates)
{
  write(x = paste(c("* **Replicate",
                    i, "**:\n"),collapse=""),file = output.file,
        append = T)
  write(x=paste(c("\t[Macs2 log](../replicates/replicate_",i,
                  "/macs_output)\n "),collapse=""), 
        file=output.file, append=T)
  write(x=paste(c("\t[Bigwig file with mapping signal](../replicates/replicate_",i,
                  "/chip_", i, ".bw )\n "),collapse=""), 
        file=output.file, append=T)
  write(x=paste(c("\t[Occupied Genomic loci in BED format](../replicates/replicate_",i,
                  "/replicate_",i,"_peaks.narrowPeak)\n "),collapse=""), 
        file=output.file, append=T)
  write(x = "\n",file=output.file, append=T)
}

if(number.replicates > 1)
{
  write(x = "\n", file=output.file, append=T)
  write(x=paste(c("The final result of this analysis was obtained combining the 
      results from the different replicates. Specifically, the", name, "genome wide 
      occupancy singal was computed as the average between the mapping signal for
      each replicate. The final set of genomic loci associated with", name,
      "was determined by intersecting the sets for each replicate. Click on the
      links below to download these final results. These files can be used as input
      in our tool **AlgaeFUN** to perform a functional annotation and determine
      the target gene set.\n"),collapse=" "), file=output.file, append=T)
  
  write(x=paste(c("* **[Genome wide Occupancy signal for ", name, " in bigWig format](./chip_average.bw)**\n "),collapse=""), 
        file=output.file, append=T)
  write(x=paste(c("* **[Genomic loci Occupied by ", name, " in BED format](./output_peaks.narrowPeak)**\n "),collapse=""), 
        file=output.file, append=T)
}