#Rscript create_Rmd.R && Rscript -e "rmarkdown::render('test.Rmd', 'pdf_document')" && Rscript -e "rmarkdown::render('test.Rmd', 'html_document')"

output.file <- "/home/fran/tmp/results/test.Rmd"
microalgae <- "Micromonas pusilla CCMP1545"
control.condition <- "t01"
experimental.condition <- "t12"
control.replicates <- 2
experimental.replicates <- 2
q.value <- 0.05
fold.change <- 2

## Document header
header <- paste0(paste0("# **Differential Gene Expression Report for ",
                        microalgae),
                 "**")

## Introduction test
write(x = header,file = output.file,append = F)
write(x = "## Experimental Design and Global Statistics",file = output.file,append = T)

intro.line <- paste(c("This report was automatically generated by **MARACAS 
                      (MicroAlgae RNA-seq and ChIP-seq Analysis)** as a 
                      differential expression analysis for the microalgae **",
                      microalgae, "**. The experimental design considers as control
                      condition **", control.condition, "** with ",
                      control.replicates, " replicates and as experimental condition **",
                      experimental.condition, "** with ", experimental.replicates,
                      " replicates. A summary of the processing of each individual
                      sample is presented below. Click on the corresponding links to 
                      download a **quality control analysis** or a file in **bigWig**
                      format containing information about the **mapping signal** for 
                      each sample:\n"
                      ), collapse="")
write(x = intro.line,file = output.file,append = T)

## Experimental design
write(x = paste(c("* Control samples **",
                  control.condition, "**:"),collapse=""),file = output.file,
      append = T)

for(i in 1:control.replicates)
{
  write(x = paste(c("\t* Control replicate **",
                    i, "**:"),collapse=""),file = output.file,
        append = T)
  write(x="[Quality Control analysis](../samples/sample_1/sample_1_fastqc.html), ", file=output.file, append=T)
  write(x="[BigWig file with mapping signal](../samples/sample_1/sample.bw)", file=output.file, append=T)
}

write(x = paste(c("* Experimental samples **",
                  experimental.condition, "**:"),collapse=""),file = output.file,
      append = T)

for(i in 1:experimental.replicates)
{
  write(x = paste(c("\t* Experimental replicate **",
                    i, "**:"),collapse=""),file = output.file,
        append = T)
  write(x="[Quality Control analysis](../samples/sample_1/sample_1_fastqc.html), ", file=output.file, append=T)
  write(x="[BigWig file with mapping signal](../samples/sample_1/sample.bw)", file=output.file, append=T)
}

## Global Gene Expression
write(x = "\n", file=output.file, append=T)
write(x="The barplot below represents the upper quantile normalized 
      global gene expression distibrution for each sample. The global gene
      expression distribution should be comparable among all samples:\n", file=output.file, append=T)

write(x = "<center>", file=output.file, append=T)
write(x = "![Normalized Global Gene Expression](./boxplot_after_normalization.png){ width=50% }", file=output.file, append=T)
write(x = "</center>", file=output.file, append=T)

## Principal components analysis
write(x = "\n", file=output.file, append=T)
write(x="A graphical representation of a **Principal Component Analysis (PCA)** 
      is presented below. Control and experimental samples should cluster into 
      different and distinct groups:\n", file=output.file, append=T)

write(x = "<center>", file=output.file, append=T)
write(x = "![Principal Components Analysis](./pca_analysis_2.png){ width=50% }", file=output.file, append=T)
write(x = "</center>", file=output.file, append=T)

## Samples Scatter plots.
write(x="Scatter plots comparing gene expression between samples is presented
      below. Samples constituting replicates for the same condition should have
      a strong diagonal signal and high correlation value.\n", file=output.file, append=T)

write(x = "<center>", file=output.file, append=T)
write(x = "![Sample Scatter Plots](./scatter_plots.png){ width=100% }", file=output.file, append=T)
write(x = "</center>", file=output.file, append=T)

write(x = "## Differential Gene Expression Analysis",file = output.file,append = T)

write(x = paste(c("In this section we present the results from a differential gene expression 
                   analysis performed using a **q-value of ", q.value, "** and 
                   a **fold-change of ", fold.change, "**. The scatter and volcano
                   plots below represent the gene differential expression. In both
                  graphs the red and blue dots represent activated and repressed
                  genes respectively. The grey dots represent genes that not 
                  differentially expressed. In the **scatter plot**, the x-coordinate
                  represents the log2 gene expression in the control condition and 
                  the y-coordinate represents the log2 gene expression in the 
                  experimental condition. Grey dots in the diagonal stand for non
                  differentially expressed genes, red dots in the upper triangle
                  represent activated genes and blue dots inthe lower triangle
                  specify repressed genes. In the **volcano plot**, the x-coordinate
                  represents the **log2 gene expression fold-change** and the 
                  y-coordinate represent the **minus log10 q-value**. Similarly,
                  grey, red and blue dots represent non differentially expressed,
                  activated and repressed genes respectively.\n 
                  " ),collapse=""),
      file = output.file,append = T)

write(x = "<center>", file=output.file, append=T)
write(x = "![Scatter Plot](./scatter_plot_control_vs_experimental.png){ width=50% }", file=output.file, append=T)
write(x = "</center>", file=output.file, append=T)

write(x = "<center>", file=output.file, append=T)
write(x = "![Scatter Plot](./volcano_plot.png){ width=50% }", file=output.file, append=T)
write(x = "</center>", file=output.file, append=T)

number.activated.genes <- 100#length(activated.genes)
number.repressed.genes <- 200#length(repressed.genes)

write(x = paste(c("Specifically, we detected **", number.activated.genes, " activated
                  genes** and **", number.repressed.genes, " repressed genes**. Click
                  on the links below to download text files containing the gene IDs for
                  activated and repressed genes. These files can be used to perform
                  a functional enrichment analysis using our tool **ALGAEFUN (
                  Microalge Functional Annotation Enrichment Analysis)**: \n
                  "),collapse=""),
      file = output.file,append = T)

write(x = "* [**Click here to download the list of activated genes.**](./activated_genes.txt)\n",
      file = output.file,append = T)
write(x = "* [**Click here to download the list of repressed genes.**](./repressed_genes.txt)\n",
      file = output.file,append = T)
